src/error.rs

```diff
@@ -1,84 +1,515 @@
+//! Contains detailed error representation.
+
 use std::error::Error;
 use std::fmt;
 use std::io;
+
 use crate::color::ExtendedColorType;
+use crate::image::ImageFormat;
 
-/// An enumeration of Image errors
+/// The generic error type for image operations.
+///
+/// This high level enum is allows, by variant matching, a rough separation of concerns between
+/// underlying IO, the caller, format specifications, and the `image` implementation.
 #[derive(Debug)]
 pub enum ImageError {
-    /// The Image is not formatted properly
-    FormatError(String),
+    /// An error was encountered while decoding.
+    ///
+    /// This means that the input data did not conform to the specification of some image format,
+    /// or that no format could be determined, or that it did not match format specific
+    /// requirements set by the caller.
+    Decoding(DecodingError),
 
-    /// The Image's dimensions are either too small or too large
-    DimensionError,
+    /// An error was encountered while encoding.
+    ///
+    /// The input image can not be encoded with the chosen format, for example because the
+    /// specification has no representation for its color space or because a necessary conversion
+    /// is ambiguous. In some cases it might also happen that the dimensions can not be used with
+    /// the format.
+    Encoding(EncodingError),
 
-    /// The Decoder does not support this image format
-    UnsupportedError(String),
+    /// An error was encountered in inputs arguments.
+    ///
+    /// This is a catch-all case for strictly internal operations such as scaling, conversions,
+    /// etc. that involve no external format specifications.
+    Parameter(ParameterError),
 
-    /// The Decoder does not support this color type
-    UnsupportedColor(ExtendedColorType),
+    /// Completing the operation would have required more resources than allowed.
+    ///
+    /// Errors of this type are limits set by the user or environment, *not* inherent in a specific
+    /// format or operation that was executed.
+    Limits(LimitError),
 
-    /// Not enough data was provided to the Decoder
-    /// to decode the image
-    NotEnoughData,
+    /// An operation can not be completed by the chosen abstraction.
+    ///
+    /// This means that it might be possible for the operation to succeed in general but
+    /// * it requires a disabled feature,
+    /// * the implementation does not yet exist, or
+    /// * no abstraction for a lower level could be found.
+    Unsupported(UnsupportedError),
 
-    /// An I/O Error occurred while decoding the image
+    /// An error occurred while interacting with the environment.
     IoError(io::Error),
+}
```

Comment:
@Nemo157 remarked on #code-review

> how big is that error type?
might be worth adding some tests like assert!(size_of::<Error>() < 24); to ensure it doesn't become too big and make copying it too slow (though I don't know what the limit at which it's worth boxing the internals is)

This points to it making a real difference in some cases: https://github.com/serde-rs/json/issues/184

Reply:
(Note: we can use the `[0][usize::from(inverted_condition)]` trick as a static assertion).

Reply:
The clippy link `large_enum_variant` fires at `200`, I'd suggest using the same or a similar limit.